File: lintcode/python/0110_minimum_path_sum.py
#!/usr/bin/python -t

# dp solution, time O(mn) space O(mn)

class Solution:
    """
    @param grid: a list of lists of integers
    @return: An integer, minimizes the sum of all numbers along its path
    """
    def minPathSum(self, grid):
        # write your code here
        # state: dp[i][j] the minimum sum at [i,j]
        # function: dp[i][j] = min(dp[i-1][j], dp[i][j-1])
        # init: dp[i][j], dp[0][0] = grid[0][0], dp[i][0], dp[0][j]
        # result: dp[m-1][n-1]
        
        m = len(grid)
        if m == 0:
            return 0
            
        n = len(grid[0])
        
        dp = [[0] * n for i in range(m)]
        
        dp[0][0] = grid[0][0]
        for i in range(1, m):
            dp[i][0] = dp[i-1][0] + grid[i][0]
            
        for j in range(1, n):
            dp[0][j] = dp[0][j-1] + grid[0][j]
            
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
                
        return dp[m-1][n-1]

File: lintcode/golang/0110_minimum_path_sum.go
package main

import "fmt"

// dp

const MaxUint = ^uint(0)
const MinUint = 0
const MaxInt = int(MaxUint >> 1)
const MinInt = -MaxInt - 1

/**
 * @param grid: a list of lists of integers
 * @return: An integer, minimizes the sum of all numbers along its path
 */
func minPathSum(grid [][]int) int {
	// write your code here
	// dp[i][j] means the minimium sum at row i, col j.
	// dp[i][j] = min(dp[i-1][j], dp[i][j-1])+ grid[i][j]
	// dp[m-1][n-1] is the answer.

	m := len(grid)
	n := len(grid[0])
	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}

	dp[0][0] = grid[0][0]

	for i := 1; i < m; i++ {
		dp[i][0] = dp[i-1][0] + grid[i][0]
	}

	for j := 1; j < n; j++ {
		dp[0][j] = dp[0][j-1] + grid[0][j]
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
		}
	}

	return dp[m-1][n-1]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	a := [][]int{
		{1, 3, 1},
		{1, 5, 1},
		{4, 2, 1},
	}
	fmt.Println(minPathSum(a))
}
